### ISSUE TITLE
Accessing model creation data inside prop initializer

### ISSUE THREAD

**Issue by dodas** [xaviergonz/mobx-keystone#275]
Hey!
I have experienced a few situations where I need to set a default property value depending on an other creation data property.
For example:
```ts
export class MyModel extends Model({
  name: prop<string | null>(null),
  // currently prop initializer does not accept any args
  nameDisplay: prop<string>((data) => data.name == null ? "(empty)" : data.name.toUpperCase()),
}) {}
```
Note the above example is contrived and could be solved with `computed`.

Could we implement such feature and pass `ModelCreationData` to the prop initializers?

Thanks!

**Comment by xaviergonz** [comment]
did you try onInit?

**Comment by xaviergonz** [comment]
or declaring a fromSnapshot in the model class

```
/**
   * Optional transformation that will be run when converting from a snapshot to the data part of the model.
   * Useful for example to do versioning and keep the data part up to date with the latest version of the model.
   *
   * @param snapshot The custom input snapshot.
   * @returns An input snapshot that must match the current model input snapshot.
   */
  fromSnapshot?(snapshot: {
    [k: string]: any
  }): SnapshotInOfObject<CreationData> & {
    [modelTypeKey]?: string
  }
```

**Comment by dodas** [comment]
Hey.
using `onInit` is my current workaround. However, one issue with that is that you would have to provide some placeholder default value in the initializer, and then override it in `onInit`.
For example:
```ts
export class MyModel extends Model({
  name: prop<string | null>(null),
  // notice the empty string here
  nameDisplay: prop<string>(""),
}) {
  onInit() {
    this.nameDisplay = this.name == null ? "(empty)" : data.name.toUpperCase();
  },
}
```
This can get out of hand quickly, if the value is something more complicated than a string.


> or declaring a fromSnapshot in the model class

`fromSnapshot` does not get called when creating instance via constructor, right? e.g. `new MyModel({ name: "my name" })`

**Comment by AndrewMorsillo** [comment]
@xaviergonz it seems like what's really needed is an equivalent to MST's `types.custom` where you can specify `name`, `fromSnapshot`, `toSnapshot`, `isTargetType`, and `getValidationMessage`. I think that might cover all the use cases for transforming/validating incoming and outgoing data? AFAICT there's no easy way right now in mobx-keystone to handle the case where data needs to be transformed on input and output such as when a database expects a specific format but you want to use a different format in the model.

Is declaring `fromSnapshot` on a model class documented anywhere? I'm a bit confused about how to use it with TS.

**Comment by xaviergonz** [comment]
> edit: as an example, Google Firestore has it's own Timestamp representation for dates. I want to represent a date from Firestore in mobx-keystone then save it back to Firestore as a Timestamp again later. With MST I simply have a types.custom which handles this for me transparently. What's the best way to do this transformation in mobx-keystone from a Timestamp to a mk type and then back to a Timestamp when creating a snapshot to save?

Right now you could try this

```
class MyModel extends Model({
  timestamp: prop(0)
}) {
  @modelAction 
  setDate(date: Date) {
    this.timestamp = +date
  }

  @computed
  get date() {
    return new Date(this.timestamp)
  }
}
```

I'll try to think if a way to make that pattern a bit more reusable

**Comment by AndrewMorsillo** [comment]
@xaviergonz Unfortunately I don't think that pattern helps much to solve the problem. That pattern wouldn't work for creation from snapshots or for persisting to snapshots as far as I can tell.

It would require creating the model then manually calling an action like setDate for _every_ field from the backend that uses a different format. When saving a snapshot to the backend it also wouldn't include the computed property in the snapshot so it would require manually cloning the snapshot then adding any of the required fields to it.

**Comment by xaviergonz** [comment]
Hmm? I'm not sure I follow

```
const m = new MyModel({ timestamp: 1234 })
m.date // Thu Jan 01 1970 01:00:01 GMT+0100
// use m.setDate to set a new date, which in the end will update the timestamp number
m.setDate(new Date(567))

const sn = getSnapshot(m)
// { timestamp: 567, ... }

const m2 = fromSnapshot(sn)
m2.date // Thu Jan 01 1970 01:00:00 GMT+0100
```

**Comment by AndrewMorsillo** [comment]
Here's a more concrete example
```
const serverData = await fetchDataFromFirestore()
serverData.someDate // this is a Firebase.FirestoreTypes.TimeStamp
// { milliseconds: number, nanoseconds: number, toDate: () => Date } ...etc
// there might be 10+ of these timestamp fields on a firestore document 

// Here's the input data problem. Some fields need to be transformed before they can be used to create a model.
const m = new MyModel(serverData) //error, no way to transform from a firestore timestamp to a number

// could manually transform the data before creating the model but that would be tedious and error prone, especially if there are more cases than just Timestamp
const serverDataTransformed = mapAllFieldsToMobxKeystoneCompatibleFields(serverData) //IMO not a great solution
const m = new MyModel(serverDataTransformed)

// Here's the output problem, some fields need to be transformed back again before being saved to the db
const sn = getSnapshot(m)
sn.someDate // 235923935 -- a number, but we need a firestore timestmp
await saveDataToFirestore(sn) //error! field someDate is a number, expected Timestamp

// Again we could manually transform the data but IMO not a great solution
const clone = JSON.parse(JSON.stringify(sn)) //to allow mutating snapshot
clone.someDate = new Firebase.Firestore.Timestamp(new Date(clone.someDate))
//....repeat for any and all fields that need to be transformed, there could be many, easy to forget one!
clone.someOtherDate = new Firebase.Firestore.Timestamp(new Date(clone.someDate))
await saveDataToFirestore(clone) //OK now but not a good solution
```

Does that clarify the issue? It would be great to have an easy way to transform incoming and outgoing data for cases where the data format of the model does not match the data format of the backend. MST custom types and snapshot transformers make this easy but I see no easy way to do it with mobx-keystone at the moment.

**Comment by xaviergonz** [comment]
just to clarify

```
const sn = getSnapshot(m)
sn.someDate // Firebase.Firestore.Timestamp -- model definition specified that field should be
```

do you expect sn.someDate to be a plain object `{seconds: x, nanoseconds: y}` or an instance of class Firebase.Firestore.Timestamp? I ask because snapshots can only hold JSON-compatible values

**Comment by AndrewMorsillo** [comment]
I think ideally there would be a way to specify that field(s) of a model should be arbitrarily transformed into any type when getting a snapshot as long as that type is compatible with the input type for model creation. 

In practice however I think I can see how that would cause difficulties. Being able to specify a transformation on snapshot for a field into any JSON compatible value would be good enough for many cases I think.

**Comment by mashaalmemon** [comment]
> or maybe your concern is that you cannot use a date as an initial prop for new?

Ultimately adding date with the `setDate` function, after instantiating the model, is not ideal because you need to do some extra work after initialization to get the object serialized properly. I have a similar issue as discussed in [Discussion 292](https://github.com/xaviergonz/mobx-keystone/discussions/292)(https://github.com/xaviergonz/mobx-keystone/discussions/292).

What would be perfect is to be able to pass in any type, as well as provide the logic to convert it to what mobx-keystone thinks is acceptable, and then have mobx-keystone do the rest without any extra work. `types.custom` in `MST` does this nicely. Would be great to have something for this.

**Comment by xaviergonz** [comment]
I'm not sure this solves the original post, but it might solve the other comments:

New property transforms + context apply/applyComputed were released as part of v0.61.0

Docs:
- https://mobx-keystone.js.org/mapsSetsDates 
- https://mobx-keystone.js.org/contexts

Let me know if that solves your case completely/partially

**Comment by xaviergonz** [comment]
Closing due to lack of feedback to the released features
